plugins/eos-blocklist/gs-plugin-eos-blocklist.c

```diff
@@ -0,0 +1,547 @@
+		g_clear_error (&local_error);
+	}
+
+	self->product_name = get_product_name (&local_error);
+
+	if (local_error != NULL) {
+		g_warning ("No system product name could be retrieved! %s", local_error->message);
+		g_clear_error (&local_error);
+	}
+
+	self->installation = flatpak_installation_new_system (cancellable, &local_error);
+	if (local_error != NULL) {
+		g_warning ("No system installation could be retrieved! %s", local_error->message);
+		g_clear_error (&local_error);
+	}
+
+	if (self->installation) {
+		self->flatpak_default_locales = flatpak_installation_get_default_locales (self->installation, &local_error);
+		if (local_error != NULL) {
+			g_warning ("No user locales could be retrieved! %s", local_error->message);
+			g_clear_error (&local_error);
+		}
+	}
+
+	g_task_return_boolean (task, TRUE);
+}
+
+static gboolean
+gs_plugin_eos_blocklist_setup_finish (GsPlugin      *plugin,
+                                      GAsyncResult  *result,
+                                      GError       **error)
+{
+	return g_task_propagate_boolean (G_TASK (result), error);
+}
+
+/* Copy of the implementation of gs_flatpak_app_get_ref_name(). */
+static const gchar *
+app_get_flatpak_ref_name (GsApp *app)
+{
+	return gs_app_get_metadata_item (app, "flatpak::RefName");
+}
+
+static gboolean
+assert_valid_locale_match (const char *app_locale,
+                           const char * const *locale_options)
+{
+	int idx;
+	g_autofree char *app_language_code = NULL;
+	g_autofree char *app_territory_code = NULL;
+	g_autofree char *app_modifier = NULL;
+
+	if (!gnome_parse_locale (app_locale, &app_language_code, &app_territory_code, NULL, &app_modifier))
+		return FALSE;
+
+	for (idx = 0; locale_options[idx] != 0; idx++) {
+		g_autofree char *opt_language_code = NULL;
+		g_autofree char *opt_territory_code = NULL;
+		g_autofree char *opt_modifier = NULL;
+
+		if (!gnome_parse_locale (locale_options[idx], &opt_language_code, &opt_territory_code, NULL, &opt_modifier))
+			return FALSE;
+
+		if (g_strcmp0 (app_language_code, opt_language_code) == 0) {
+			/* If the main is a match, try to match territory eg. US or GB */
+			if (app_territory_code != NULL && opt_territory_code != NULL) {
+				if (g_strcmp0 (app_territory_code, opt_territory_code) == 0) {
+					/* If the territory is a match, try to match the modifier eg. latin/cyrillic */
+					if (app_modifier != NULL && opt_modifier != NULL) {
+						if (g_strcmp0 (app_modifier, opt_modifier) == 0)
+							return TRUE;
+					} else {
+						/* If the [modifier] of the app_locale or from the locale options
+						 * is not defined, this is a desireable app */
+						return TRUE;
+					}
+				}
+			} else {
+				/* If the [territory] of the app_locale or from the locale options
+				 * is not defined, this is a desireable app */
+				return TRUE;
+			}
+		}
+	}
+
+	return FALSE;
+}
+
+static gboolean
+locale_is_compatible (GsPluginEosBlocklist *self,
```

Comment:
I wonder if we could take a different approach here:

1. Remove all our unmaintained locale-specific knowledge apps from https://github.com/endlessm/gnome-software-data/blob/master/app-info/eos-extra.txt so that they don't appear in the carousel or Editor's Choice
2. Then just remove this entire plugin and allow 10 different encyclopedias in 10 different languages to appear

Maybe that's a separate discussion.

Reply:
https://phabricator.endlessm.com/T33450 is the separate discussion I guess. OK let's carry this forwards.